File: shared/typescript/graders.ts
/**
 * Code-based graders (lesson 14). Pure, deterministic, free to run.
 *
 * Anthropic's evaluation guidance ranks graders in this order:
 *   code-based   - fast, deterministic, preferred wherever it is possible
 *   model-based  - for nuanced judgements, with an explicit rubric
 *   human        - flexible, slow, expensive
 *
 * Reach for a model judge only for what code cannot check.
 */
import type { GradeResult } from "./eval.ts";

/** Every listed field matches the expected value exactly. */
export function exactFields<TOutput, TExpected = TOutput>(
  fields: readonly (keyof TOutput & keyof TExpected & string)[],
): (output: TOutput, expected: TExpected) => GradeResult {
  return (output, expected) => {
    const wrong = fields.filter(
      (field) => output[field] !== (expected[field] as unknown),
    );
    return wrong.length === 0
      ? { score: 1, reason: "all fields match" }
      : {
          score: 1 - wrong.length / fields.length,
          reason: wrong
            .map(
              (field) =>
                `${field}: got ${JSON.stringify(output[field])}, want ${JSON.stringify(expected[field])}`,
            )
            .join(", "),
        };
  };
}

/** The text parses as JSON. */
export function isValidJson(text: string): GradeResult {
  try {
    JSON.parse(text);
    return { score: 1, reason: "parses" };
  } catch (error) {
    return {
      score: 0,
      reason: `not JSON: ${error instanceof Error ? error.message : String(error)}`,
    };
  }
}

/** The text matches a regular expression. */
export function matchesPattern(
  pattern: RegExp,
  label = "pattern",
): (output: string) => GradeResult {
  return (output) =>
    pattern.test(output)
      ? { score: 1, reason: `${label} matched` }
      : { score: 0, reason: `${label} did not match` };
}

/** Word count falls inside an inclusive range. */
export function withinWordCount(
  min: number,
  max: number,
): (output: string) => GradeResult {
  return (output) => {
    const words = output.trim().split(/\s+/).filter(Boolean).length;
    return words >= min && words <= max
      ? { score: 1, reason: `${words} words` }
      : { score: 0, reason: `${words} words, wanted ${min}-${max}` };
  };
}

/** None of the listed strings appears in the output (case-insensitive). */
export function avoidsPhrases(
  phrases: readonly string[],
): (output: string) => GradeResult {
  return (output) => {
    const lower = output.toLowerCase();
    const found = phrases.filter((phrase) => lower.includes(phrase.toLowerCase()));
    return found.length === 0
      ? { score: 1, reason: "no banned phrases" }
      : { score: 0, reason: `contains: ${found.join(", ")}` };
  };
}

File: shared/python/course/graders.py
"""Code-based graders (lesson 14). Pure, deterministic, free to run.

Anthropic's evaluation guidance ranks graders in this order:
    code-based   - fast, deterministic, preferred wherever it is possible
    model-based  - for nuanced judgements, with an explicit rubric
    human        - flexible, slow, expensive

Reach for a model judge only for what code cannot check.
"""

from __future__ import annotations

import ast
import json
import re
from collections.abc import Callable, Sequence
from typing import Any

from course.evals import GradeResult


def exact_fields(fields: Sequence[str]) -> Callable[..., GradeResult]:
    """Every listed field matches the expected value exactly."""

    def grade(output: dict[str, Any], expected: dict[str, Any], *_: Any) -> GradeResult:
        wrong = [f for f in fields if output.get(f) != expected.get(f)]
        if not wrong:
            return GradeResult(1.0, "all fields match")
        detail = ", ".join(
            f"{f}: got {output.get(f)!r}, want {expected.get(f)!r}" for f in wrong
        )
        return GradeResult(1 - len(wrong) / len(fields), detail)

    return grade


def is_valid_json(output: str, *_: Any) -> GradeResult:
    """The text parses as JSON."""
    try:
        json.loads(output)
    except (json.JSONDecodeError, TypeError) as error:
        return GradeResult(0.0, f"not JSON: {error}")
    return GradeResult(1.0, "parses")


def is_valid_python(output: str, *_: Any) -> GradeResult:
    """The text parses as Python. Syntax only - it is not executed."""
    try:
        ast.parse(output)
    except SyntaxError as error:
        return GradeResult(0.0, f"SyntaxError: {error.msg}")
    return GradeResult(1.0, "parses")


def matches_pattern(pattern: str, label: str = "pattern") -> Callable[..., GradeResult]:
    """The text matches a regular expression."""
    compiled = re.compile(pattern)

    def grade(output: str, *_: Any) -> GradeResult:
        if compiled.search(output):
            return GradeResult(1.0, f"{label} matched")
        return GradeResult(0.0, f"{label} did not match")

    return grade


def within_word_count(minimum: int, maximum: int) -> Callable[..., GradeResult]:
    """Word count falls inside an inclusive range."""

    def grade(output: str, *_: Any) -> GradeResult:
        words = len(output.split())
        if minimum <= words <= maximum:
            return GradeResult(1.0, f"{words} words")
        return GradeResult(0.0, f"{words} words, wanted {minimum}-{maximum}")

    return grade


def avoids_phrases(phrases: Sequence[str]) -> Callable[..., GradeResult]:
    """None of the listed strings appears in the output (case-insensitive)."""

    def grade(output: str, *_: Any) -> GradeResult:
        lower = output.lower()
        found = [phrase for phrase in phrases if phrase.lower() in lower]
        if not found:
            return GradeResult(1.0, "no banned phrases")
        return GradeResult(0.0, f"contains: {', '.join(found)}")

    return grade
